### ISSUE TITLE
[feature request] Hooks for phase tracking and triggers.

### ISSUE THREAD

**Issue by roya0045** [keras-team/keras#7701]
Greetings,

I'm using Keras with TF as backend, I have implemented a custom metric to log the data. I know that callback have some "hooks" on the end of a batch or an epoch. I was wondering if something similar could be done with the distinct phases of a neural network [learning, validating and testing/evaluating], having such hooks would allow to trigger either distinct metrics on each phase or only use a metric on some other phases.

This would improve metric flexibility and potentially allow more feature to be implemented. I think it  could be implemented quite easily by adding a variable that keep the state of the model( though I am not familiar enough with the whole library and backends.)

Thanks
